<?xml version="1.0"?>
<launch>
    <!-- ╔══════════════════════════════════════════════════════════╗ -->
    <!-- ║  直接在这里修改算法和地图，然后 roslaunch 即可           ║ -->
    <!-- ║  Algorithms: forex_nav | frtree | cogniplan | rrtx      ║ -->
    <!-- ║  Maps: office1  office2  office3  complex_office        ║ -->
    <!-- ║        maze_simple  maze  tunnel1  cmu                  ║ -->
    <!-- ║        library  hkust1  hkust2  hkust3                  ║ -->
    <!-- ╚══════════════════════════════════════════════════════════╝ -->
    <arg name="algorithm" default="rrtx" />
    <arg name="map_name"  default="hkust1"   />

    <!-- ==================== 其他全局参数 ==================== -->
    <arg name="rviz"      default="true" />
    <arg name="robot_type" default="zsl-1" doc="Robot type: zsl-1 | go2" />
    <arg name="max_vel"   default="2.0" />
    <arg name="max_acc"   default="2.0" />
    <arg name="planning_height" default="0.4" />

    <!-- ==================== Per-Map Configurations ==================== -->

    <!-- office1 -->
    <!-- MAP_TRAJECTORIES[office1]="-6 10 4.6 -13.8 5.4 13.3 -6.5 -14" -->
    <arg name="office1_init_x"    value="-6"/>
    <arg name="office1_init_y"    value="10"/>
    <arg name="office1_init_yaw"  value="-1.57"/>
    <arg name="office1_goal_x"   value="5.20"/>
    <arg name="office1_goal_y"   value="-9.32"/>
    <arg name="office1_box_min_x" value="-10.0"/>
    <arg name="office1_box_min_y" value="-15.0"/>
    <arg name="office1_box_max_x" value="10.0"/>
    <arg name="office1_box_max_y" value="15.0"/>
    <arg name="office1_rotate_z"  value="0.0"/>
    <arg name="office1_trans_x"   value="0.0"/>
    <arg name="office1_trans_y"   value="0.0"/>
    <arg name="office1_pcd"       value="office.pcd"/>

    <!-- office2 -->
    <!-- MAP_TRAJECTORIES[office2]="8.5 -8.5 -8.65 8.8 -0.88 -8.5 8.7 9.0" -->
    <arg name="office2_init_x"    value="8.5"/>
    <arg name="office2_init_y"    value="-8.5"/>
    <arg name="office2_init_yaw"  value="2.1"/>
    <arg name="office2_goal_x"   value="0.00"/>
    <arg name="office2_goal_y"   value="0.00"/>
    <arg name="office2_box_min_x" value="-10.0"/>
    <arg name="office2_box_min_y" value="-15.0"/>
    <arg name="office2_box_max_x" value="10.0"/>
    <arg name="office2_box_max_y" value="15.0"/>
    <arg name="office2_rotate_z"  value="0.0"/>
    <arg name="office2_trans_x"   value="0.0"/>
    <arg name="office2_trans_y"   value="0.0"/>
    <arg name="office2_pcd"       value="office2.pcd"/>

    <!-- office3 -->
    <!-- MAP_TRAJECTORIES[office3]="8.5 -12.5 -7.6 12.8 -1.35 -12.8 8.6 13" -->
    <arg name="office3_init_x"    value="8.5"/>
    <arg name="office3_init_y"    value="-12.5"/>
    <arg name="office3_init_yaw"  value="2.1"/>
    <arg name="office3_goal_x"   value="0.00"/>
    <arg name="office3_goal_y"   value="0.00"/>
    <arg name="office3_box_min_x" value="-10.0"/>
    <arg name="office3_box_min_y" value="-15.0"/>
    <arg name="office3_box_max_x" value="10.0"/>
    <arg name="office3_box_max_y" value="15.0"/>
    <arg name="office3_rotate_z"  value="0.0"/>
    <arg name="office3_trans_x"   value="0.0"/>
    <arg name="office3_trans_y"   value="0.0"/>
    <arg name="office3_pcd"       value="office3.pcd"/>

    <!-- complex_office -->
    <!-- MAP_TRAJECTORIES[complex_office]="14 -14 -13 13 -13 -14 14.5 14" -->
    <arg name="complex_office_init_x"    value="14.0"/>
    <arg name="complex_office_init_y"    value="13.0"/>
    <arg name="complex_office_init_yaw"  value="3.14"/>
    <arg name="complex_office_goal_x"   value="0.00"/>
    <arg name="complex_office_goal_y"   value="0.00"/>
    <arg name="complex_office_box_min_x" value="-17.0"/>
    <arg name="complex_office_box_min_y" value="-17.0"/>
    <arg name="complex_office_box_max_x" value="17.0"/>
    <arg name="complex_office_box_max_y" value="17.0"/>
    <arg name="complex_office_rotate_z"  value="0.0"/>
    <arg name="complex_office_trans_x"   value="-15.0"/>
    <arg name="complex_office_trans_y"   value="15.0"/>
    <arg name="complex_office_pcd"       value="complex_office_rt.pcd"/>

    <!-- maze_simple -->
    <!-- MAP_TRAJECTORIES[maze_simple]="8.3 -19 -8.8 19 -2.5 -19 8.5 19.6" -->
    <arg name="maze_simple_init_x"    value="8.3"/>
    <arg name="maze_simple_init_y"    value="-19"/>
    <arg name="maze_simple_init_yaw"  value="2.1"/>
    <arg name="maze_simple_goal_x"   value="0.00"/>
    <arg name="maze_simple_goal_y"   value="0.00"/>
    <arg name="maze_simple_box_min_x" value="-10.0"/>
    <arg name="maze_simple_box_min_y" value="-15.0"/>
    <arg name="maze_simple_box_max_x" value="10.0"/>
    <arg name="maze_simple_box_max_y" value="15.0"/>
    <arg name="maze_simple_rotate_z"  value="0.0"/>
    <arg name="maze_simple_trans_x"   value="0.0"/>
    <arg name="maze_simple_trans_y"   value="0.0"/>
    <arg name="maze_simple_pcd"       value="maze_simple.pcd"/>

    <!-- maze -->
    <arg name="maze_init_x"    value="14"/>
    <arg name="maze_init_y"    value="13.75"/>
    <arg name="maze_init_yaw"  value="3.14"/>
    <arg name="maze_goal_x"   value="0.00"/>
    <arg name="maze_goal_y"   value="0.00"/>
    <arg name="maze_box_min_x" value="-17.5"/>
    <arg name="maze_box_min_y" value="-17.5"/>
    <arg name="maze_box_max_x" value="17.5"/>
    <arg name="maze_box_max_y" value="17.5"/>
    <arg name="maze_rotate_z"  value="0.0"/>
    <arg name="maze_trans_x"   value="-17.5"/>
    <arg name="maze_trans_y"   value="17.5"/>
    <arg name="maze_pcd"       value="maze_rt.pcd"/>

    <!-- tunnel1 -->
    <!-- MAP_TRAJECTORIES[tunnel1]="4.0 -18.5 -8.8 14.3 3.7 -19.4 8.2 19.9" -->
    <arg name="tunnel1_init_x"    value="4.0"/>
    <arg name="tunnel1_init_y"    value="-18.5"/>
    <arg name="tunnel1_init_yaw"  value="1.57"/>
    <arg name="tunnel1_goal_x"   value="0.00"/>
    <arg name="tunnel1_goal_y"   value="0.00"/>
    <arg name="tunnel1_box_min_x" value="-12.0"/>
    <arg name="tunnel1_box_min_y" value="-22.0"/>
    <arg name="tunnel1_box_max_x" value="12.0"/>
    <arg name="tunnel1_box_max_y" value="22.0"/>
    <arg name="tunnel1_rotate_z"  value="180.0"/>
    <arg name="tunnel1_trans_x"   value="-7.0"/>
    <arg name="tunnel1_trans_y"   value="0.0"/>
    <arg name="tunnel1_pcd"       value="tunnel_cave_trans.pcd"/>

    <!-- cmu -->
    <!-- MAP_TRAJECTORIES[cmu]="-22.4 -41.8 40.7 22.8 6.4 -45.4 39.4 23" -->
    <arg name="cmu_init_x"    value="-22.4"/>
    <arg name="cmu_init_y"    value="-41.8"/>
    <arg name="cmu_init_yaw"  value="1.57"/>
    <arg name="cmu_goal_x"   value="0.00"/>
    <arg name="cmu_goal_y"   value="0.00"/>
    <arg name="cmu_box_min_x" value="-65.0"/>
    <arg name="cmu_box_min_y" value="-50.0"/>
    <arg name="cmu_box_max_x" value="65.0"/>
    <arg name="cmu_box_max_y" value="50.0"/>
    <arg name="cmu_rotate_z"  value="0.0"/>
    <arg name="cmu_trans_x"   value="-65.0"/>
    <arg name="cmu_trans_y"   value="-18.5"/>
    <arg name="cmu_pcd"       value="indoor_cmu.pcd"/>

    <!-- library -->
    <arg name="library_init_x"    value="9.8"/>
    <arg name="library_init_y"    value="13.6"/>
    <arg name="library_init_yaw"  value="3.14"/>
    <arg name="library_goal_x"   value="0.00"/>
    <arg name="library_goal_y"   value="0.00"/>
    <arg name="library_box_min_x" value="-13.0"/>
    <arg name="library_box_min_y" value="-20.0"/>
    <arg name="library_box_max_x" value="14.0"/>
    <arg name="library_box_max_y" value="20.0"/>
    <arg name="library_rotate_z"  value="0.0"/>
    <arg name="library_trans_x"   value="5.0"/>
    <arg name="library_trans_y"   value="11.0"/>
    <arg name="library_pcd"       value="LibraryLG2.pcd"/>

    <!-- hkust1 -->
    <!-- MAP_TRAJECTORIES[hkust1]="-8 -26.2 -1.2 30 18.5 -25.2 20.1 25" -->
    <arg name="hkust1_init_x"    value="16.54"/>
    <arg name="hkust1_init_y"    value="25.52"/>
    <arg name="hkust1_init_yaw"  value="0.00"/>
    <arg name="hkust1_goal_x"   value="-3.44"/>
    <arg name="hkust1_goal_y"   value="-30.31"/>
    <arg name="hkust1_box_min_x" value="-25.0"/>
    <arg name="hkust1_box_min_y" value="-33.0"/>
    <arg name="hkust1_box_max_x" value="25.0"/>
    <arg name="hkust1_box_max_y" value="33.0"/>
    <arg name="hkust1_rotate_z"  value="0.0"/>
    <arg name="hkust1_trans_x"   value="-102.0"/>
    <arg name="hkust1_trans_y"   value="-11.0"/>
    <arg name="hkust1_pcd"       value="hkust_1.pcd"/>

    <!-- hkust2 -->
    <!-- MAP_TRAJECTORIES[hkust2]="7.3 -27.6 19.5 28.5 -19 -27 -3.3 26" -->
    <arg name="hkust2_init_x"    value="7.3"/>
    <arg name="hkust2_init_y"    value="-27.6"/>
    <arg name="hkust2_init_yaw"  value="1.57"/>
    <arg name="hkust2_goal_x"   value="0.00"/>
    <arg name="hkust2_goal_y"   value="0.00"/>
    <arg name="hkust2_box_min_x" value="-25.0"/>
    <arg name="hkust2_box_min_y" value="-33.0"/>
    <arg name="hkust2_box_max_x" value="25.0"/>
    <arg name="hkust2_box_max_y" value="33.0"/>
    <arg name="hkust2_rotate_z"  value="0.0"/>
    <arg name="hkust2_trans_x"   value="-172.0"/>
    <arg name="hkust2_trans_y"   value="-11.0"/>
    <arg name="hkust2_pcd"       value="hkust_2.pcd"/>

    <!-- hkust3 -->
    <arg name="hkust3_init_x"    value="-1.0"/>
    <arg name="hkust3_init_y"    value="12.0"/>
    <arg name="hkust3_init_yaw"  value="0.0"/>
    <arg name="hkust3_goal_x"   value="0.00"/>
    <arg name="hkust3_goal_y"   value="0.00"/>
    <arg name="hkust3_box_min_x" value="-25.0"/>
    <arg name="hkust3_box_min_y" value="-33.0"/>
    <arg name="hkust3_box_max_x" value="25.0"/>
    <arg name="hkust3_box_max_y" value="33.0"/>
    <arg name="hkust3_rotate_z"  value="0.0"/>
    <arg name="hkust3_trans_x"   value="-172.0"/>
    <arg name="hkust3_trans_y"   value="-11.0"/>
    <arg name="hkust3_pcd"       value="hkust_3.pcd"/>

    <!-- ==================== Resolve Map Parameters ==================== -->
    <!-- Allow override of init_x, init_y, init_yaw; otherwise use map-specific defaults -->
    <arg name="init_x"   default="$(eval arg(arg('map_name') + '_init_x'))"/>
    <arg name="init_y"   default="$(eval arg(arg('map_name') + '_init_y'))"/>
    <arg name="init_yaw" default="$(eval arg(arg('map_name') + '_init_yaw'))"/>

    <!-- Preset goal (resolved from map_name; set via set_benchmark_points.launch) -->
    <arg name="goal_x" default="$(eval arg(arg('map_name') + '_goal_x'))"/>
    <arg name="goal_y" default="$(eval arg(arg('map_name') + '_goal_y'))"/>

    <!-- Box bounds (resolved from map_name) -->
    <arg name="box_min_x" value="$(eval arg(arg('map_name') + '_box_min_x'))"/>
    <arg name="box_min_y" value="$(eval arg(arg('map_name') + '_box_min_y'))"/>
    <arg name="box_max_x" value="$(eval arg(arg('map_name') + '_box_max_x'))"/>
    <arg name="box_max_y" value="$(eval arg(arg('map_name') + '_box_max_y'))"/>

    <!-- PCD file and map transforms (resolved from map_name) -->
    <arg name="pcd_file" value="$(eval arg(arg('map_name') + '_pcd'))"/>
    <arg name="rotate_z" value="$(eval arg(arg('map_name') + '_rotate_z'))"/>
    <arg name="trans_x"  value="$(eval arg(arg('map_name') + '_trans_x'))"/>
    <arg name="trans_y"  value="$(eval arg(arg('map_name') + '_trans_y'))"/>
    <!-- trans_z: library uses 3.0, all others use -0.3 -->
    <arg name="trans_z"  value="$(eval 3.0 if arg('map_name') == 'library' else -0.3)"/>

    <!-- Map size for local sensing: derived from box bounds with margin -->
    <arg name="map_x_size" value="$(eval arg('box_max_x') - arg('box_min_x') + 4.0)"/>
    <arg name="map_y_size" value="$(eval arg('box_max_y') - arg('box_min_y') + 4.0)"/>

    <!-- ==================== Algorithm: forex_nav ==================== -->
    <include file="$(find nav_exp_env)/base_env/launch/forex_nav_legged.launch"
             if="$(eval arg('algorithm') == 'forex_nav')">
        <arg name="max_vel" value="$(arg max_vel)" />
        <arg name="max_acc" value="$(arg max_acc)" />
        <arg name="map_x_size" value="$(arg map_x_size)" />
        <arg name="map_y_size" value="$(arg map_y_size)" />
        <arg name="planning_height" value="$(arg planning_height)" />
        <arg name="robot_type" value="$(arg robot_type)" />
        <arg name="rviz" value="$(arg rviz)" />
        <arg name="init_x" value="$(arg init_x)" />
        <arg name="init_y" value="$(arg init_y)" />
        <arg name="init_yaw" value="$(arg init_yaw)" />
        <arg name="pcd_file" value="$(arg pcd_file)" />
        <arg name="rotate_z" value="$(arg rotate_z)" />
        <arg name="trans_x" value="$(arg trans_x)" />
        <arg name="trans_y" value="$(arg trans_y)" />
        <arg name="trans_z" value="$(arg trans_z)" />
    </include>

    <!-- ==================== Algorithm: frtree ==================== -->
    <include file="$(find nav_exp_env)/frtree/launch/frtree_nav_legged.launch"
             if="$(eval arg('algorithm') == 'frtree')">
        <arg name="max_vel" value="$(arg max_vel)" />
        <arg name="max_acc" value="$(arg max_acc)" />
        <arg name="map_x_size" value="$(arg map_x_size)" />
        <arg name="map_y_size" value="$(arg map_y_size)" />
        <arg name="planning_height" value="$(arg planning_height)" />
        <arg name="robot_type" value="$(arg robot_type)" />
        <arg name="init_x" value="$(arg init_x)" />
        <arg name="init_y" value="$(arg init_y)" />
        <arg name="init_yaw" value="$(arg init_yaw)" />
        <arg name="pcd_file" value="$(arg pcd_file)" />
        <arg name="rotate_z" value="$(arg rotate_z)" />
        <arg name="trans_x" value="$(arg trans_x)" />
        <arg name="trans_y" value="$(arg trans_y)" />
        <arg name="trans_z" value="$(arg trans_z)" />
    </include>

    <!-- ==================== Algorithm: cogniplan ==================== -->
    <include file="$(find nav_exp_env)/cogniplan/launch/cogniplan_nav_legged.launch"
             if="$(eval arg('algorithm') == 'cogniplan')">
        <arg name="max_vel" value="$(arg max_vel)" />
        <arg name="max_acc" value="$(arg max_acc)" />
        <arg name="map_x_size" value="$(arg map_x_size)" />
        <arg name="map_y_size" value="$(arg map_y_size)" />
        <arg name="planning_height" value="$(arg planning_height)" />
        <arg name="robot_type" value="$(arg robot_type)" />
        <arg name="rviz" value="$(arg rviz)" />
        <arg name="init_x" value="$(arg init_x)" />
        <arg name="init_y" value="$(arg init_y)" />
        <arg name="init_yaw" value="$(arg init_yaw)" />
        <arg name="pcd_file" value="$(arg pcd_file)" />
        <arg name="rotate_z" value="$(arg rotate_z)" />
        <arg name="trans_x" value="$(arg trans_x)" />
        <arg name="trans_y" value="$(arg trans_y)" />
        <arg name="trans_z" value="$(arg trans_z)" />
    </include>

    <!-- ==================== Algorithm: rrtx ==================== -->
    <include file="$(find nav_exp_env)/rrtx/launch/rrtx_nav_legged.launch"
             if="$(eval arg('algorithm') == 'rrtx')">
        <arg name="max_vel" value="$(arg max_vel)" />
        <arg name="max_acc" value="$(arg max_acc)" />
        <arg name="map_x_size" value="$(arg map_x_size)" />
        <arg name="map_y_size" value="$(arg map_y_size)" />
        <arg name="planning_height" value="$(arg planning_height)" />
        <arg name="robot_type" value="$(arg robot_type)" />
        <arg name="rviz" value="$(arg rviz)" />
        <arg name="init_x" value="$(arg init_x)" />
        <arg name="init_y" value="$(arg init_y)" />
        <arg name="init_yaw" value="$(arg init_yaw)" />
        <arg name="pcd_file" value="$(arg pcd_file)" />
        <arg name="rotate_z" value="$(arg rotate_z)" />
        <arg name="trans_x" value="$(arg trans_x)" />
        <arg name="trans_y" value="$(arg trans_y)" />
        <arg name="trans_z" value="$(arg trans_z)" />
    </include>

    <!-- ==================== Preset Goal Sender ==================== -->
    <!-- Click "Publish Point" in rviz to send preset goal to /move_base_simple/goal -->
    <!-- "2D Nav Goal" still works independently (different rviz tool, same topic) -->
    <arg name="use_preset_goal" default="true" />
    <node pkg="nav_exp_env" type="preset_goal_sender.py" name="preset_goal_sender" output="screen"
          if="$(arg use_preset_goal)">
        <param name="goal_x" value="$(arg goal_x)" />
        <param name="goal_y" value="$(arg goal_y)" />
        <param name="planning_height" value="$(arg planning_height)" />
    </node>

    <!-- ==================== Benchmark Metrics Recorder ==================== -->
    <node pkg="nav_exp_env" type="benchmark_recorder.py" name="benchmark_recorder" output="screen">
        <param name="algorithm" value="$(arg algorithm)" />
        <param name="map_name"  value="$(arg map_name)" />
        <param name="goal_reach_threshold" value="0.2" />
    </node>

</launch>
